core/store/src/trie/iterator.rs

```diff
@@ -130,6 +135,9 @@ impl<'a> TrieIterator<'a> {
                         hash = *child.unwrap_hash();
                         *status = CrumbStatus::At;
                         self.key_nibbles.extend(existing_key.iter());
+                        if key.is_empty() {
+                            *prefix_boundary = true;
+                        }
```

Comment:
Two things here.  Firstly, we want to set the prefix_boundary flag only if we were called from `seek_prefix`.  `seek_nibble_slice` needs `is_prefix_seek: bool` argument or alternatively we can add `seek_nibble_slice_impl` method with the implementation and have `seek_prefix` and `seek_nibble_slice` call that.

Secondly, we want to set the flag on all but the last node I think.  This way we’ll visit the node that we’ve reached but then as soon as we exit it we’ll exit all the other nodes.  This should be doable with the following code structure:

```rust
        let mut prev_prefix_boundary = &mut false;
        loop {
            *prev_prefix_boundary = true;
            self.descend_into_node(&hash)?;
            let Crumb { status, node, prefix_boundary } = self.trail.last_mut().unwrap();
            prev_prefix_boundary = prefix_boundary;
            match &node.node { /* ... */ }
        }
    }
```